### PR TITLE
fix(controller): Randomly expr expression fail to resolve

### LINKED ISSUE #6257: When condition randomly invalid
## Summary 

Running the following workflow template it fails sometimes. Most of the time it is valid and running without any issues.
```yaml
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: test-skip
  generateName: test-skip-
  namespace: argo
spec:
  entrypoint: diamond
  arguments:
    parameters:
      - name: jobsToExecute
  templates:
    - name: diamond
      dag:
        tasks:
          - name: A
            template: echo
            when: "{{=sprig.has('A', sprig.splitList('|', sprig.trim(workflow.parameters.jobsToExecute)))}}"
            arguments:
              parameters: [{name: message, value: A}]
          - name: B
            template: echo
            when: "{{=sprig.has('B', sprig.splitList('|', sprig.trim(workflow.parameters.jobsToExecute)))}}"
            arguments:
              parameters: [{name: message, value: B}]
            dependencies:
              - A
          - name: C
            template: echo
            when: "{{=sprig.has('C', sprig.splitList('|', sprig.trim(workflow.parameters.jobsToExecute)))}}"
            arguments:
              parameters: [{name: message, value: C}]
            dependencies:
              - A
          - name: D
            template: echo
            when: "{{=sprig.has('D', sprig.splitList('|', sprig.trim(workflow.parameters.jobsToExecute)))}}"
            arguments:
              parameters: [{name: message, value: D}]
            dependencies:
              - B
              - C

    - name: echo
      inputs:
        parameters:
          - name: message
      container:
        image: alpine:3.7
        command: [echo, "{{inputs.parameters.message}}"]
```

The entered argument values are:
- `A|B`
- `B|C`
- `C|D`
- ...

Sometimes we get the following error `Invalid 'when' expression '{{=sprig.has('C', sprig.splitList('|', sprig.trim(workflow.parameters.jobsToExecute)))}}': Invalid token: '{{=' (hint: try wrapping the affected expression in quotes ("))`. If we start a new workflow from this template with the same argument it is a valid expression and there are no issues. We expect that there is no invalid expression exception every time.

## Diagnostics

What Kubernetes provider are you using? GKE

What version of Argo Workflows are you running? 3.1.1

What executor are you running? Docker

Did this work in a previous version? I.e. is it a regression? First try, no information about previous versions

```bash
# Logs from the workflow controller:
time="2021-06-30T12:42:20.554Z" level=info msg="Processing workflow" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:20.568Z" level=info msg="Updated phase  -> Running" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:20.570Z" level=info msg="DAG node test-skip-mbmq6 initialized Running" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:20.571Z" level=info msg="All of node test-skip-mbmq6.A dependencies [] completed" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:20.572Z" level=info msg="Pod node test-skip-mbmq6-765059746 initialized Pending" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:20.587Z" level=info msg="Created pod: test-skip-mbmq6.A (test-skip-mbmq6-765059746)" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:20.601Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=75592506 workflow=test-skip-mbmq6
time="2021-06-30T12:42:30.588Z" level=info msg="Processing workflow" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:30.588Z" level=info msg="Updating node test-skip-mbmq6-765059746 exit code 0" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:30.588Z" level=info msg="Setting node test-skip-mbmq6-765059746 outputs: {\"artifacts\":[{\"name\":\"main-logs\",\"gcs\":{\"key\":\"2021/06/30/test-skip-mbmq6/test-skip-mbmq6-765059746/main.log\"}}]}" namespace=argo workflow
test-skip-mbmq6
time="2021-06-30T12:42:30.588Z" level=info msg="Updating node test-skip-mbmq6-765059746 status Pending -> Succeeded" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:30.589Z" level=info msg="All of node test-skip-mbmq6.B dependencies [A] completed" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:30.591Z" level=info msg="Pod node test-skip-mbmq6-748282127 initialized Pending" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:30.611Z" level=info msg="Created pod: test-skip-mbmq6.B (test-skip-mbmq6-748282127)" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:30.612Z" level=info msg="Skipped node test-skip-mbmq6-731504508 initialized Error (message: Invalid 'when' expression '{{=sprig.has('C', sprig.splitList('|', sprig.trim(workflow.parameters.jobsToExecute)))}}': Invalid token:
'{{=' (hint: try wrapping the affected expression in quotes (\")))" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:30.625Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=75592579 workflow=test-skip-mbmq6
time="2021-06-30T12:42:30.633Z" level=info msg="cleaning up pod" action=labelPodCompleted key=argo/test-skip-mbmq6-765059746/labelPodCompleted
time="2021-06-30T12:42:40.612Z" level=info msg="Processing workflow" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:40.613Z" level=info msg="Updating node test-skip-mbmq6-748282127 exit code 0" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:40.613Z" level=info msg="Setting node test-skip-mbmq6-748282127 outputs: {\"artifacts\":[{\"name\":\"main-logs\",\"gcs\":{\"key\":\"2021/06/30/test-skip-mbmq6/test-skip-mbmq6-748282127/main.log\"}}]}" namespace=argo workflow
test-skip-mbmq6
time="2021-06-30T12:42:40.613Z" level=info msg="Updating node test-skip-mbmq6-748282127 status Pending -> Succeeded" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:40.615Z" level=info msg="Skipped node test-skip-mbmq6-714726889 initialized Omitted (message: omitted: depends condition not met)" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:40.615Z" level=info msg="Outbound nodes of test-skip-mbmq6 set to [test-skip-mbmq6-714726889]" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:40.615Z" level=info msg="node test-skip-mbmq6 phase Running -> Error" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:40.615Z" level=info msg="node test-skip-mbmq6 finished: 2021-06-30 12:42:40.615612009 +0000 UTC" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:40.615Z" level=info msg="Checking daemoned children of test-skip-mbmq6" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:40.615Z" level=info msg="Updated phase Running -> Error" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:40.615Z" level=info msg="Marking workflow completed" namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:40.615Z" level=info msg="Checking daemoned children of " namespace=argo workflow=test-skip-mbmq6
time="2021-06-30T12:42:40.625Z" level=info msg="Workflow update successful" namespace=argo phase=Error resourceVersion=75592654 workflow=test-skip-mbmq6
time="2021-06-30T12:42:40.635Z" level=info msg="cleaning up pod" action=labelPodCompleted key=argo/test-skip-mbmq6-748282127/labelPodCompleted
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by sarabala1979**
Signed-off-by: Saravanan Balasubramanian <sarabala1979@gmail.com>
closes #6257 
Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

Tips:

* Your PR needs to pass the required checks before it can be approved. If the check is not required (e.g. E2E tests) it does not need to pass
* Sign-off your commits to pass the DCO check: `git commit --signoff`.
* Run `make pre-commit -B` to fix codegen or lint problems. 
* Say how how you tested your changes. If you changed the UI, attach screenshots.

### CODE REVIEW

File: workflow/common/common.go [hunk 1]
```diff
@@ -139,7 +139,7 @@ const (
 	// GlobalVarWorkflowDuration is the current duration of this workflow
 	GlobalVarWorkflowDuration = "workflow.duration"
 	// GlobalVarWorkflowParameters is a JSON string containing all workflow parameters
-	GlobalVarWorkflowParameters = "workflow.parameters"
+	GlobalVarWorkflowParameters = "workflow.parameters.json"
```
Comment (sarabala1979) [inline workflow/common/common.go:142]: Randomly `bellows.Expand` failed to expend  all global variable `workflow.parameters.xxx` if `workflow.parameters` expended first.  We should have a different name for `workflow.parameters` json format.

Comment (alexec) [reply workflow/common/common.go:142]: we should raise a bug with bellows?

Comment (alexec) [reply workflow/common/common.go:142]: maybe not, I think it does not make sense - but can we make sure that this is not used - new in v3.1?

Comment (alexec) [reply workflow/common/common.go:142]: also - should work for fasttemplate

Comment (sarabala1979) [reply workflow/common/common.go:142]: `fasttemplate` will just string replace. it will work. I can ask the community channel for this variable.

Comment (alexec) [reply workflow/common/common.go:142]: if it is in v3.0, then changing it would be breaking change

**Comment by alexec** [comment]
feat: Expose workflow.paramteres with JSON string of all params (#2341)

Date to 2020, so can't change name.

File: workflow/common/common.go (hunk 1 above)
Comment (alexec) [reply workflow/common/common.go:142]: I think we should exclude this from `{{=` templates. Users should not need it.

Comment (sarabala1979) [reply workflow/common/common.go:142]: we can prevent to have {{= in `when` element. So this is the way when will have only `govaluate` format.

**Review by alexec — APPROVED** [review]

**Pull request merged**